collectors/slabinfo.plugin/slabinfo.c

```diff
@@ -0,0 +1,389 @@
+
+	static procfile *ff = NULL;
+
+	if(unlikely(!ff)) {
+		ff = procfile_reopen(ff, PLUGIN_SLABINFO_PROCFILE, " ,:" , PROCFILE_FLAG_DEFAULT);
+		if(unlikely(!ff)) {
+			error("<- Cannot open file '%s", PLUGIN_SLABINFO_PROCFILE);
+			exit(1);
+		}
+	}
+
+	ff = procfile_readall(ff);
+	if(unlikely(!ff)) {
+		error("<- Cannot read file '%s'", PLUGIN_SLABINFO_PROCFILE);
+		exit(0);
+	}
+
+
+	// Iterate on all lines to populate / update the slabinfo struct
+	size_t lines = procfile_lines(ff), l;
+
+	debug(D_PLUGIN_SLABINFO, "   Read %lu lines from procfile", lines);
+	for(l = 2; l < lines; l++) {
+		if (unlikely(procfile_linewords(ff, l) < 14)) {
+			debug(D_PLUGIN_SLABINFO, "    Line %lu has only %lu words, skipping", l, procfile_linewords(ff,l));
+			continue;
+		}
+
+		char *name = procfile_lineword(ff, l, 0);
+		struct slabinfo *s = get_slabstruct(name);
+
+		s->active_objs    = str2uint64_t(procfile_lineword(ff, l, 1));
+		s->num_objs       = str2uint64_t(procfile_lineword(ff, l, 2));
+		s->obj_size       = str2uint64_t(procfile_lineword(ff, l, 3));
+		s->obj_per_slab   = str2uint64_t(procfile_lineword(ff, l, 4));
+		s->pages_per_slab = str2uint64_t(procfile_lineword(ff, l, 5));
+
+		s->tune_limit     = str2uint64_t(procfile_lineword(ff, l, 7));
+		s->tune_batchcnt  = str2uint64_t(procfile_lineword(ff, l, 8));
+		s->tune_shared_factor = str2uint64_t(procfile_lineword(ff, l, 9));
+
+		s->data_active_slabs = str2uint64_t(procfile_lineword(ff, l, 11));
+		s->data_num_slabs    = str2uint64_t(procfile_lineword(ff, l, 12));
+		s->data_shared_avail = str2uint64_t(procfile_lineword(ff, l, 13));
+
+		uint32_t memperslab = s->pages_per_slab * SLAB_PAGE_SIZE;
+		// Internal fragmentation: loss per slab, due to objects not being a multiple of pagesize
+		//uint32_t lossperslab = memperslab - s->obj_per_slab * s->obj_size;
+
+		// Total usage = slabs * pages per slab * page size
+		s->mem_usage = (uint64_t)(s->data_num_slabs * memperslab);
+
+		// Wasted memory (filling): slabs allocated but not filled: sum total slab - sum total objects
+		s->mem_waste = s->mem_usage - (uint64_t)(s->active_objs * s->obj_size);
+		//if (s->data_num_slabs > 1)
+		//	s->mem_waste += s->data_num_slabs * lossperslab;
+
+
+		// Slab filling efficiency
+		if (s->num_objs > 0)
+			s->obj_filling = 100 * s->active_objs / s->num_objs;
+		else
+			s->obj_filling = 0;
+
+		debug(D_PLUGIN_SLABINFO, "    Updated slab %s: %lu %lu %lu %lu %lu / %lu %lu %lu / %lu %lu %lu / %lu %lu %hhu",
+			name, s->active_objs, s->num_objs, s->obj_size, s->obj_per_slab, s->pages_per_slab,
+			s->tune_limit, s->tune_batchcnt, s->tune_shared_factor,
+			s->data_active_slabs, s->data_num_slabs, s->data_shared_avail,
+			s->mem_usage, s->mem_waste, s->obj_filling);
+	}
+
+	return slabinfo_root;
+}
+
+
+
+unsigned int do_slab_stats(int update_every) {
+
+	static unsigned int loops = 0;
+	struct slabinfo *sactive = NULL, *s = NULL;
+
+	// Main processing loop
+	while ( running ) {
```

Comment:
Extra spaces.

Reply:
Fixed. Is there any linter that could check this automatically ?

Reply:
#6677 is on its way to being merged.